### DISCUSSION TITLE
Building for a very lightweight platform?

### DISCUSSION THREAD

**Discussion by seamusdemora** [Arkq/bluez-alsa#695] [Q&A] [answered]
Hello again,
Back with another question:  I want to try `bluez-alsa` on a tiny little [Raspberry Pi 2W](https://www.raspberrypi.com/products/raspberry-pi-zero-2-w/). I'm currently running the "Lite" (headless) version of the RPi OS 'bookworm' (64-bit). I could run the 'bullseye' version of the OS if that would "improve my chances" of success. Any thoughts or opinions on that? 

<s>If I do the build on another platform, is there an install procedure (I assume that there are config files that get set up during the build)??</s> (answered in the installation README.md)

<s>Is there a list of packages I must have to build/'make' `bluez-alsa`?</s> (answered in the wiki) 

Thanks for your help!

**Comment by borine** [comment, marked as answer]
I run bluealsa on an *even more tiny* raspberry pi zero (mark1, 32bit processor) using bookworm, and previously with bullseye, buster, and stretch. So my opinion is no problem (depending on exactly you are expecting it to do!)

**Reply by seamusdemora** [reply]
Did you build on the pi zero - or did you build elsewhere & "import"?

**Reply by borine** [reply]
I have my own custom-built cross-compile environment. But you certainly can build on a Pi zero, it will just take a long time. Or you could build on a pi4 or pi5 which should be reasonably quick.

**Reply by seamusdemora** [reply]
As for my expectations, I *think* they're reasonable... I want to play a "rainfall soundtrack" using `mpg123` (or maybe the in-built player if it has the features I need) & hear the audio through a single "JBL Flip 5" BT speaker. Also considering the addition of a tiny "rocker switch" connected to GPIO for volume control. 

I'm *currently running* `pipewire` & `wireplumber` on the zero 2W. It's been working *reasonably well*, but it does just stop playing (audio stops) sometimes. I've not been able to get a handle on the cause, so I'm going to try something else: `bluezalsa`

**Reply by seamusdemora** [reply]
I will ***guess*** that this should be moved to a new Q... if not, pls advise.

OK, I'm in `Step 3` and have a bunch more questions: 
<s>1. Can you explain how to use the "`configure` command line" to: `--enable-mpg123` ?</s> answered in wiki
2. Does `mpg123` support **also** require `--enable-mp3lame`? 
3. Do I need any other A2DP codecs to play mp3 files through a BT (JBL Flip 5) speaker?
4. I *guess* I will also need the `--enable-cli` to use `bluealsa-cli` tool as I have no GUI ??
5. Will the `--enable-rfcomm` option enable RPi-to-RPi serial comms? 
6. I think I'd like to use *some of* the `systemd` options, so will add `--enable-systemd` option - BUT what about others? Can you make a recommendation? 
7. Do I need `fdk-aac`?? (I hope not as it sounds *complicated*)

**Reply by borine** [reply]
You do not need any additional options to simply play mp3 files to a bluetooth speaker. Keep your first build simple:
```
../configure --enable-systemd
make
```
--enable-rfcomm is only for HFP or HSP telephone voice profiles - not relevant to your use case.

do not try any other codecs until you have the basics working. SBC should be fine for your use case, and by far the most reliable choice. You probably will not need `bluealsa-cli`. Just:
* install bluealsa,
* enable and start the bluealsa daemon in systemd (see the relevant wiki page),
*  pair and connect the speaker (see the pairing wiki page)
* run `mpg123 -o alsa -a bluealsa music_file.mp3`

**Reply by seamusdemora** [reply]
"Keep it simple". :)  I like the sound of that. lol

Just two more questions, and then I'll try to leave you alone for a while: 
Does it make any difference whether I am running on a 64-bit OS - or a 32-bit OS?  
Does it make any difference whether I am running `bullseye` vs `bookworm`? 
Sorry for the waffling, but I just this morning got some new information on the RPi Zero 2W that might cause me to move **back** to 32-bit 'bullseye'!

**Reply by borine** [reply]
> Does it make any difference whether I am running on a 64-bit OS - or a 32-bit OS?

No, bluealsa works fine as a 32-bit or 64-bit application, so long as it is compiled for its host system.

> Does it make any difference whether I am running bullseye vs bookworm?

No, I've used it without issues on both.

**Reply by seamusdemora** [reply]
### THE ISSUE DESCRIBED BELOW HAS BEEN RESOLVED, BUT SEE QUESTION AT END OF REPLY

Uh-oh...  what did I do wrong? 
```bash
$ mpg123 -o alsa -a bluealsa rainstorm.mp3
High Performance MPEG 1.0/2.0/2.5 Audio Player for Layers 1, 2 and 3
	version 1.31.2; written and copyright by Michael Hipp and others
	free software (LGPL) without any warranty but with best wishes
ALSA lib ../../../src/asound/bluealsa-pcm.c:1477:(_snd_pcm_bluealsa_open) Couldn't get BlueALSA PCM: PCM not found
[src/libout123/modules/alsa.c:open_alsa():181] error: cannot open device bluealsa
[src/libout123/libout123.c:check_output_module():947] error: Module 'alsa' device open failed.
[src/libout123/libout123.c:out123_open():439] error: Found no driver out of [alsa] working with device bluealsa.
main: [src/mpg123.c:check_fatal_output():334] error: out123 error 3: failure loading driver module
```

checked systemd bluealsa:
```bash
$ systemctl status bluealsa.service
● bluealsa.service - BlueALSA service
     Loaded: loaded (/lib/systemd/system/bluealsa.service; enabled; preset: enabled)
     Active: active (running) since Wed 2024-04-17 20:34:08 UTC; 6min ago
       Docs: man:bluealsa(8)
   Main PID: 531 (bluealsa)
      Tasks: 4 (limit: 380)
        CPU: 376ms
     CGroup: /system.slice/bluealsa.service
             └─531 /usr/bin/bluealsa -S -p a2dp-source -p a2dp-sink

Apr 17 20:34:08 rpi2w systemd[1]: Starting bluealsa.service - BlueALSA service...
Apr 17 20:34:08 rpi2w systemd[1]: Started bluealsa.service - BlueALSA service
```

checked pairing & trusting:
```bash
$ bluetoothctl
Agent registered
[CHG] Controller B8:27:EB:DE:B9:12 Pairable: yes
[bluetooth]# power on
Changing power on succeeded
[bluetooth]# devices
Device B8:F6:53:9B:1A:97 JBL Flip 5
[bluetooth]# devices Paired
Device B8:F6:53:9B:1A:97 JBL Flip 5
[bluetooth]# devices Trusted
Device B8:F6:53:9B:1A:97 JBL Flip 5
```

Tried again after "connecting" in `bluetoothctl`: 
```bash
[bluetooth]# connect B8:F6:53:9B:1A:97
Attempting to connect to B8:F6:53:9B:1A:97
[CHG] Device B8:F6:53:9B:1A:97 Connected: yes
[NEW] Endpoint /org/bluez/hci0/dev_B8_F6_53_9B_1A_97/sep1
[NEW] Endpoint /org/bluez/hci0/dev_B8_F6_53_9B_1A_97/sep2
[NEW] Transport /org/bluez/hci0/dev_B8_F6_53_9B_1A_97/sep1/fd0
Connection successful
[NEW] Player /org/bluez/hci0/dev_B8_F6_53_9B_1A_97/player0 [default]
[CHG] Device B8:F6:53:9B:1A:97 ServicesResolved: yes
[CHG] Transport /org/bluez/hci0/dev_B8_F6_53_9B_1A_97/sep1/fd0 Volume: 0x0010 (16)
[JBL Flip 5]# quit
```

And now... SUCCESS!
```bash
$ nohup mpg123 -o alsa -a bluealsa --loop -1 rainstorm.mp3 &
[1] 835

```

Will this `connect` w/ `bluetoothctl` always be required to play music??

**Reply by borine** [reply]
> Will this connect w/ bluetoothctl always be required to play music??

Some speakers auto connect to a known device when the speaker is  turned on, others do not. I've never used a JBL Flip so I do  not know about yours. If your Pi was booted and bluealsa running before the JBL was powered on, and the JBL did not connect when it was turned on, then I guess it does not auto connect.

You can connect with a one-liner as
```
bluetoothctl connect B8:F6:53:9B:1A:97
```

**Reply by graham8** [reply]
My Flip 3 reconnects just fine.

**Reply by borine** [reply]
> My Flip 3 reconnects just fine.

thanks @graham8 : Then I guess the Flip 5 probably does try to connect also. Two possible reasons why it might fail in this case:
* bluealsa was not running at the time that the Flip was turned on. Generally devices only try once on power up, they do not repeat the attempt if it fails.
* The Flip connected to some other device that it is paired with - possibly a nearby mobile phone.

**Comment by borine** [comment]
I don't like the idea of using `nohup`. That command is for starting long-running applications from a login shell - but surely you don't really intend to have to login to start this thing?

I've put together a simple non-interactive solution as a demonstration, based on `systemd` and using `bluealsa-cli`, that is active as soon as the pi is booted, and automatically plays the rainstorm sound whenever a bluetooth speaker is connected.  (I know I previously suggested you would not need `bluealsa-cli`, so apologies for that). I've avoided hard-coding the speaker bt address so you can use it with any speaker. Hopefully this will give you a starting point from which to develop; you can modify it to suit your precise requirements.

If you are not using the "pi" account, then change the path to the `rainstorm.mp3` file in the bash script mpg123 command, and the `User=` directive in the systemd service file. Make sure the chosen user is a member of the `audio` group in order to be able to use bluetooth audio.

 
Create the following 2 files:

**/usr/local/bin/rainstorm**
```
#!/bin/bash
exec &>/dev/null
for device in $(/usr/bin/bluetoothctl devices Paired | /usr/bin/sed -n 's/^Device \([^ ]*\).*$/\1/p')
do
	/usr/bin/bluealsa-cli list-pcms | grep 'a2dpsrc/sink$' && break
	if /usr/bin/bluetoothctl info "$device" | /usr/bin/grep -i 0000110b ; then
		/usr/bin/bluetoothctl connect "$device"
	fi
done
while true; do
	/usr/bin/mpg123 -o alsa -a bluealsa --quiet --no-control --loop -1 /home/pi/rainstorm.mp3
	/usr/bin/grep -q -m 1 '^PCMAdded.*a2dp.*sink' <(/usr/bin/bluealsa-cli monitor)
done
```
Make the rainstorm script executable
```
sudo chmod a+x /usr/local/bin/rainstorm
```

**/usr/local/lib/systemd/system/rainstorm.service**
```
[Unit]
Description=Rainstorm audio service
Requires=bluealsa.service
After=bluealsa.service

[Service]
Type=simple
User=pi
Group=audio
ExecStart=/usr/local/bin/rainstorm

[Install]
WantedBy=bluetooth.target
```

Start the new service with:
```
sudo systemctl daemon-reload
sudo systemctl enable --now rainstorm.service
```

The rainstorm sound will play as soon as a bluetooth speaker is connected, and run continuously until the speaker is disconnected. It will start every time the pi is booted, so no login is necessary.

**Reply by seamusdemora** [reply]
Yes... using `nohup` was the only way I could get `mpg123` to "cooperate" by running in the background. And I really didn't intend to login to start it... my plan was (once I got everything ["ironed out"](https://idioms.thefreedictionary.com/iron+out)) to start `mpg123` from a `cron` job: start it at 11PM & "kill" it at 8AM. 

Running `mpg123` continuously in the background was my approach to getting things "ironed out"; testing for anything that halted execution & interrupted the sound.  While evaluating `pipewire` I noticed that it would sometimes run for days, but other times, on other platforms, it stopped randomly (or it seemed so to me). I never found what caused this. 

And speaking of "random stops", when I came in this morning to see how things were going w/ `bluez-alsa` & `mpg123`, the music had stopped. Here's what I found:

1. `ps -e | grep mpg123` showed it to still be running, but "disassociated" from the SSH `pts` I started it under; i.e.:
```bash
$ ps -e | grep mpg123
712 ?        00:04:04 mpg123
```   

2. checked `bluealsa`...  this looks "healthy" to me: 
```bash
$ systemctl status bluealsa.service
● bluealsa.service - BlueALSA service
     Loaded: loaded (/lib/systemd/system/bluealsa.service; enabled; preset: enabled)
     Active: active (running) since Thu 2024-04-18 05:59:08 UTC; 6h ago
       Docs: man:bluealsa(8)
   Main PID: 529 (bluealsa)
      Tasks: 6 (limit: 380)
        CPU: 2min 10.432s
     CGroup: /system.slice/bluealsa.service
             └─529 /usr/bin/bluealsa -S -p a2dp-source -p a2dp-sink

Apr 18 05:59:07 rpi2w systemd[1]: Starting bluealsa.service - BlueALSA service...
Apr 18 05:59:08 rpi2w systemd[1]: Started bluealsa.service - BlueALSA service.
```
   
3. checked w/ `bluetoothctl`... this also looked "healthy": 
```bash
$ bluetoothctl
[JBL Flip 5]# power on
Changing power on succeeded
[JBL Flip 5]# devices Paired
Device B8:F6:53:9B:1A:97 JBL Flip 5
[JBL Flip 5]# devices Trusted
Device B8:F6:53:9B:1A:97 JBL Flip 5
[JBL Flip 5]# devices Connected
Device B8:F6:53:9B:1A:97 JBL Flip 5
[JBL Flip 5]# exit
```

4. but when I tried to re-start `mpg123`... does not look "healthy":
```bash
$ mpg123 -o alsa -a bluealsa rainstorm.mp3
High Performance MPEG 1.0/2.0/2.5 Audio Player for Layers 1, 2 and 3
	version 1.31.2; written and copyright by Michael Hipp and others
	free software (LGPL) without any warranty but with best wishes
ALSA lib ../../../src/asound/bluealsa-pcm.c:1477:(_snd_pcm_bluealsa_open) Couldn't get BlueALSA PCM: PCM not found
[src/libout123/modules/alsa.c:open_alsa():181] error: cannot open device bluealsa
[src/libout123/libout123.c:check_output_module():947] error: Module 'alsa' device open failed.
[src/libout123/libout123.c:out123_open():439] error: Found no driver out of [alsa] working with device bluealsa.
main: [src/mpg123.c:check_fatal_output():334] error: out123 error 3: failure loading driver module
```  

And so, I'm unsure what caused this. A `reboot` brought everything "back to normal/good"; including an "automatic connect" between my RPi Zero 2W and the JBL Flip 5. Re-started `mpg123` & it's still playing...  

A coupla' things: 
1. I am **not** wedded to `mpg123`... if you think there's a better/more reliable player I'm happy to try it. 
2. Read some things last night that suggest perhaps `64-bit bookworm` may not be "optimal" for RPi Zero 2W. I'll drop back to either `32-bit bookworm` or `64-bit bullseye` once I get a clear answer to a question I've posted at the RPi site. 
3. I'm also going to try your `systemd` suggestion... I'm not proficient w/ `systemd`, but I think it will accommodate the "start at 11PM & stop at 8AM" objective. And re-building to add `bluealsa-cli` is not a problem... the first build went quite well, and didn't even take that long :)  

Any other guidance is muchly appreciated

**Comment by seamusdemora** [comment]
> The rainstorm sound will play as soon as a bluetooth speaker is connected, and run continuously until the speaker is disconnected. It will start every time the pi is booted, so no login is necessary.

Very cool!  

I've rebuilt w/ the `bluealsa-cli` option & copied/pasted your script & unit file in the correct locations. Scheduling the start and stop for playing the rainstorm track looks like something I *might* be able to handle w/ my meager systemd skills (using [Option B from here](https://askubuntu.com/a/1480977/831935)). 

In the meantime, I'm running your script from `cron` - b/c it's faster/easier. Here's what I've added to the `root crontab` to start playing at 10:30 PM, and stop playing at 7:30 AM (note I keep my RPis sync'd to UTC) : 

```
30 3 * * * /usr/local/bin/rainstorm >> /home/pi/rainstormlog.txt 2>&1
30 12 * * * /usr/bin/pkill rainstorm && /usr/bin/pkill mpg123 >> /home/pi/rainstormlog.txt 2>&1
```

**Reply by seamusdemora** [reply]
@borine 

I am still having the occasional issue when playback stops (speaker goes silent). I am running from the `cron` job described above in the previous post - instead of the `systemd` unit you suggested. Again, I did this b/c I wanted to be able to play only during "sleep hours". 

Sound stopped again last night (don't know exactly when... I was asleep :)  .  I did a "post mortem" on the only things I could think of that might give a clue (see below). I wonder if it might help to turn "ON" some of the output from `mpg123` & `rainstorm`, and capture that in a logfile?  Below, I've made a couple of minor changes to the `rainstorm` script to give it a "voice" - any analysis or comments? Can you think of other things I should try? Should I turn this back over to `systemd`, and come up with a way to stop it?

```bash
pi@rpi2w:~ $ systemctl status bluealsa.service
● bluealsa.service - BlueALSA service
     Loaded: loaded (/lib/systemd/system/bluealsa.service; enabled; preset: enabled)
     Active: active (running) since Fri 2024-04-19 18:08:58 UTC; 2 days ago
       Docs: man:bluealsa(8)
   Main PID: 518 (bluealsa)
      Tasks: 6 (limit: 380)
        CPU: 50min 20.624s
     CGroup: /system.slice/bluealsa.service
             └─518 /usr/bin/bluealsa -S -p a2dp-source -p a2dp-sink

Apr 19 18:08:57 rpi2w systemd[1]: Starting bluealsa.service - BlueALSA service...
Apr 19 18:08:58 rpi2w bluealsa[518]: Couldn't get HCI version: Network is down
Apr 19 18:08:58 rpi2w systemd[1]: Started bluealsa.service - BlueALSA service.
pi@rpi2w:~ $ ps -e | grep rainstorm
11814 ?        00:00:00 rainstorm
pi@rpi2w:~ $ ps -e | grep mpg123
11820 ?        05:26:59 mpg123
```  

Minor revisions to `rainstorm`: 

```bash
#!/bin/bash
# exec &>/dev/null              # suppress all output
for device in $(/usr/bin/bluetoothctl devices Paired | /usr/bin/sed -n 's/^Device \([^ ]*\).*$/\1/p')
do
        /usr/bin/bluealsa-cli list-pcms | grep 'a2dpsrc/sink$' && break
        if /usr/bin/bluetoothctl info "$device" | /usr/bin/grep -i 0000110b ; then
                /usr/bin/bluetoothctl connect "$device"
        fi
done
while true; do
#       /usr/bin/mpg123 -o alsa -a bluealsa --quiet --no-control --loop -1 /home/pi/rainstorm.mp3
        /usr/bin/mpg123 -o alsa -a bluealsa --no-control --loop -1 /home/pi/rainstorm.mp3
#       /usr/bin/grep -q -m 1 '^PCMAdded.*a2dp.*sink' <(/usr/bin/bluealsa-cli monitor)
        /usr/bin/grep -m 1 '^PCMAdded.*a2dp.*sink' <(/usr/bin/bluealsa-cli monitor)
done
```

**Reply by borine** [reply]
I really think that this failure has nothing to do with bluealsa or mpg123; there is some system event at play here, and so the answer is more likely to be found in the system logs. Perhaps a power-saving mode is kicking in? I know this can be problem for wifi with the RPiOS lite image - I put the following in `/etc/rc.local` 
```
iw wlan0 set power_save off
```
but I do not know if this makes any difference to bluetooth.

**Reply by seamusdemora** [reply]
hmm... Thanks - I'll check this out & let you know what I find. 

Would it make sense in this case to use `bluetoothctl` to check if the speaker is connected? i.e. something like:

```bash
if [ bluetoothctl devices Connected != "Device B8:F6:53:9B:1A:97 JBL Flip 5" ] ; then 
    < attempt to connect >
fi
```

**Reply by borine** [reply]
Your shell syntax needs some work, but in principle you could do that if you prefer. But beware that `bluetoothctl` reports a device as connected so long as at least one service is connected, so in general it cannot guarantee that A2DP is connected. In your case that may not matter; perhaps the JBL will disconnect any other services if A2DP fails, I can't be sure because I don't have one. My aim was to avoid any specific bluetooth address, and allow the script to use any A2DP speaker that happens to be connected. Of course that may be a completely wrong approach for your scenario. Also the script was written without any time limits in mind, so it may well be possible to make changes to optimize it for use with systemd timers or cron. I'm not offering free consultancy for that though :smile:

**Reply by seamusdemora** [reply]
I'm seeing a new "condition" since yesterday in the systemd status for `bluealsa`:
```
$ systemctl status bluealsa.service
● bluealsa.service - BlueALSA service
     Loaded: loaded (/lib/systemd/system/bluealsa.service; enabled; preset: enabled)
     Active: active (running) since Tue 2024-04-23 02:12:23 UTC; 25min ago
       Docs: man:bluealsa(8)
   Main PID: 526 (bluealsa)
      Tasks: 5 (limit: 380)
        CPU: 397ms
     CGroup: /system.slice/bluealsa.service
             └─526 /usr/bin/bluealsa -S -p a2dp-source -p a2dp-sink

Apr 23 02:12:22 rpi2w systemd[1]: Starting bluealsa.service - BlueALSA service...
Apr 23 02:12:23 rpi2w systemd[1]: Started bluealsa.service - BlueALSA service.
Apr 23 02:12:23 rpi2w bluealsa[526]: Couldn't get HCI version: Network is down
```  

The last line: "Network is down". This was run shortly after a reboot, and during the time when the music is not playing. AFAIK, I've made no changes to my system... any idea what's causing this?

**Reply by arkq** [reply]
> Couldn't get HCI version: Network is down

In most cases this warning log is not harmful. It might pop out in cases when BlueALSA service starts before the HCI is UP. In your case it will do absolutely nothing. However, it might affect cases when one is using mSBC with HFP. Then, BlueALSA might not be able to determine whether local HCI interface can support mSBC.
